### PR TITLE
Move setup.cfg to pyproject.toml

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This pull request moves all the package configuration options from `setup.cfg` to `pyproject.toml` to bring `astropy` inline with [PEP621](https://peps.python.org/pep-0621/).

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/coordinates/tests/test_representation_arithmetic.py [hunk 1]
```diff
@@ -1407,7 +1407,7 @@ def test_operation_unit_spherical(self, diff_cls, op, args):
         # and division there will be sign flips in the spherical distance.
         expected_c = op(
             rep.represent_as(CartesianRepresentation, {"s": CartesianDifferential}),
-            *args
+            *args,
```
Comment (astrofrog) [inline astropy/coordinates/tests/test_representation_arithmetic.py:1410]: Can you make these changes as part of a separate PR?

Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_representation_arithmetic.py:1410]: Unfortunately, black decided these changes were necessary after I made these changes.

I am investigating why this might be.

Comment (WilliamJamieson) [reply astropy/coordinates/tests/test_representation_arithmetic.py:1410]: Further investigation has revealed to me that by just migrating the `[metadata]` entries from `setup.cfg` to `[project]` in the `pyproject.toml` seems to trigger `black` wanting to inject all these commas.

I have determined that the changes made by `black` triggered by this change are totally fine under without the `pyproject.toml` migration. Thus I think the simplest course of action is to merge those changes first and then merge the meaningful changes proposed here.

File: docs/conf.py [hunk 2]
```diff
@@ -126,10 +131,11 @@
 
 extensions += ["sphinx_changelog", "sphinx_design"]
 
-# Grab minversion from setup.cfg
-setup_cfg = configparser.ConfigParser()
-setup_cfg.read(os.path.join(os.path.pardir, "setup.cfg"))
-__minimum_python_version__ = setup_cfg["options"]["python_requires"].replace(">=", "")
+# Grab minversion from pyproject.toml
+with (Path(__file__).parent.parent / "pyproject.toml").open("rb") as f:
```
Comment (nstarman) [inline docs/conf.py:135]:
```suggestion
with (Path(__file__).parents[1] / "pyproject.toml").open("rb") as f:
```

File: pyproject.toml [hunk 3]
```diff
@@ -6,9 +125,89 @@ requires = ["setuptools",
...
+namespaces = false
+
+[tool.setuptools.package-data]
+"*" = [
+    "data/*",
+    "data/*/*",
+    "data/*/*/*",
+    "data/*/*/*/*",
+    "data/*/*/*/*/*",
+    "data/*/*/*/*/*/*",
```
Comment (nstarman) [inline pyproject.toml:143]: I know this was a direct transfer to pyproject, but perhaps this can be cleaned up?

Comment (nstarman) [reply pyproject.toml:143]: Also fine to push to a future PR

**Review by nstarman — COMMENTED** [review]
🫰

File: pyproject.toml (hunk 3 above)
Comment (WilliamJamieson) [reply pyproject.toml:143]: I agree! However, this is attempting do as much as possible a 1-to-1 port to the `pyproject.toml` file.

So I think this can wait until a future PR.

**Review by nstarman — APPROVED** [review]
Thanks @WilliamJamieson!  It’s great to migrate away from legacy config files. Aside from the changelog check not passing, let’s get this merged!

**Comment by pllim** [comment]
Would be nice if @astropy/astropy-project-release-team also get a chance to review. Let's wait a bit more. Thanks!

File: docs/changes/15247.other.rst
Comment (pllim) [inline docs/changes/15247.other.rst:1]: Do we really need a change log for this? User won't see any difference.

**Comment by nstarman** [comment]
Also, squash merge?

File: docs/changes/15247.other.rst
Comment (WilliamJamieson) [reply docs/changes/15247.other.rst:1]: IMOP it should get a change log simply because its quite a big change in the sense that we are re-defining all our core project metadata.

**Comment by pllim** [comment]
> Initiated cron tests to double check pyproject installation

Good idea! Better be safe than sorry.

File: pyproject.toml [hunk 4]
```diff
@@ -1,3 +1,122 @@
...
+    "coverage[toml]",
+    "skyfield>=1.20",
+    "sgp4>=2.3",
+]
+recommended = [
+    "scipy>=1.5",
+    "matplotlib>=3.3,!=3.4.0,!=3.5.2",
+]
+all = [
+    "astropy[recommended]",  # installs the [recommended] dependencies
```
Comment (astrofrog) [inline pyproject.toml:70]: So does this use 'recommended' from this pyproject file or does it actually get astropy from PyPI and get recommended from there?

File: pyproject.toml [hunk 5]
```diff
@@ -1,3 +1,122 @@
...
+    "bottleneck",
+    "ipython>=4.2",
+    "pytest>=7.0",
+    "typing_extensions>=3.10.0.1",
+    "fsspec[http]>=2022.8.2",
+    "s3fs>=2022.8.2",
+    "pre-commit",
+]
+docs = [
+    "astropy[recommended]",  # installs the [recommended] dependencies
```
Comment (astrofrog) [inline pyproject.toml:93]: This is a change from before right? Is it really needed for the docs to build?

Comment (WilliamJamieson) [reply pyproject.toml:93]: Before we exactly repeated the recommended requirements for the docs requirements. That's why I just directly referenced them instead.

**Review by astrofrog — APPROVED** [review]

**Review by pllim — CHANGES_REQUESTED** [review]
Wait, we also need to update this: https://github.com/astropy/astropy/blob/main/.github/workflows/update_astropy_iers_data_pin.py

**Comment by pllim** [comment]
Maybe it is best to grep for "setup.cfg" references in the code base to ensure we didn't miss anything else. From the root dir:

```
git clean -xdf
grep -r "setup\.cfg" .
```

**Comment by saimn** [comment]
I started working on a similar branch a while ago (https://github.com/astropy/astropy/compare/main...saimn:astropy:pyproject?expand=1) but didn't go further because I thought we would want to first make sure that our pytest plugins actually support this.
IIRC it is fine, the problematic one being pytest-doctestplus, I think `doctest_norecursedirs` and `doctest_subpackage_requires` were correctly parsed, but I wanted first to add tests with `pyproject.toml` in pytest-doctestplus, which did not happen...
So just saying, would be better to do that first instead of finding in a few months that some configuration isn't used as expected.

Also I think the package-data section can be removed and we can rely on setuptools-sm instead (https://github.com/astropy/astropy/commit/8bdf61e4ff959acb59b1c0c6dbdb8d0574f6a19a).

File: docs/changes/15247.other.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Switch from using ``setup.cfg`` for project configuration to using ``pyproject.toml``
```
Comment (astrofrog) [inline docs/changes/15247.other.rst:1]: Repetition of 'project configuration'

Comment (WilliamJamieson) [reply docs/changes/15247.other.rst]: Done

File: .github/workflows/update_astropy_iers_data_pin.yml [hunk 7]
```diff
@@ -36,7 +36,7 @@ jobs:
       run: |
         git config user.name github-actions
         git config user.email github-actions@github.com
-        git add setup.cfg
+        git add pypr
```
Comment (pllim) [inline .github/workflows/update_astropy_iers_data_pin.yml:39]: This seems incomplete.

File: .github/workflows/update_astropy_iers_data_pin.py [hunk 8]
```diff
@@ -7,24 +7,25 @@
...
 
 for iline in range(len(lines)):
     if "astropy-iers-data" in lines[iline]:
         changed_lines += 1
-        lines[iline] = f"    astropy-iers-data>={last_version}\n"
+        lines[iline] = f'    "astropy-iers-data>={last_version}",\n'
 
 if changed_lines == 0:
-    print("No line found containing astropy-iers-data in setup.cfg")
+    print("No line found containing astropy-iers-data in pyproject.toml")
```
Comment (pllim) [inline .github/workflows/update_astropy_iers_data_pin.py:23]: Why not reuse the `project_file` in prints?

File: pyproject.toml [hunk 9]
```diff
@@ -6,9 +125,89 @@ requires = ["setuptools",
...
+[tool.setuptools.package-data]
+"*" = [
+    "data/*",
+    "data/*/*",
+    "data/*/*/*",
+    "data/*/*/*/*",
+    "data/*/*/*/*/*",
+    "data/*/*/*/*/*/*",
+]
+'astropy' = ['astropy.cfg', 'CITATION']
```
Comment (pllim) [inline pyproject.toml:145]: Why use single quotes here and not double quotes? The latter is more consistent with the rest of the file.

Comment (WilliamJamieson) [reply pyproject.toml:145]: Because I wasn't consistent. I can setup an auto formatter for it?

Comment (pllim) [reply pyproject.toml:145]: Hmm I am surprised I caught this before that pre-commit thingy.

Comment (WilliamJamieson) [reply pyproject.toml:145]: The pre-commit doesn't check this file for that sort of thing

File: docs/development/ccython.rst [hunk 10]
```diff
@@ -52,14 +52,14 @@ you probably want to install its header files along side the Python module.
...
        file to include those header files in the package. For example, the
        `astropy.wcs` package has the following entries in the
-       ``[options.package_data]`` section::
+       ``[tool.setuptools.package_data]`` section::
 
-           [options.package_data]
+           [tool.setuptools.package_data]
            ...
-           astropy.wcs = include/*/*.h
+           'astropy.wcs' = ["include/*/*.h"]
```
Comment (pllim) [inline docs/development/ccython.rst:62]:
```suggestion
           "astropy.wcs" = ["include/*/*.h"]
```

File: docs/development/testguide.rst [hunk 11]
```diff
@@ -27,17 +27,18 @@ dependency itself, in addition to several ``pytest`` plugins that are used by
...
+file. Developers who want to run the test suite will need to either
 install pytest-astropy directly::
 
     pip install pytest-astropy
 
 or install the core package in 'editable' mode specifying the ``[test]``
 option::
 
-    pip install -e .[test]
+    pip install -e ".[test]"
```
Comment (pllim) [inline docs/development/testguide.rst:41]: Are the quotes really necessary here?

File: pyproject.toml [hunk 12]
```diff
@@ -1,14 +1,213 @@
...
+fitscheck = "astropy.io.fits.scripts.fitscheck:main"
+fitsdiff = "astropy.io.fits.scripts.fitsdiff:main"
+fitsheader = "astropy.io.fits.scripts.fitsheader:main"
+fitsinfo = "astropy.io.fits.scripts.fitsinfo:main"
+samp_hub = "astropy.samp.hub_script:hub_script"
+showtable = "astropy.table.scripts.showtable:main"
+volint = "astropy.io.votable.volint:main"
+wcslint = "astropy.wcs.wcslint:main"
+
+
```
Comment (pllim) [inline pyproject.toml:119]:
```suggestion
```

**Review by pllim — COMMENTED** [review]
@saimn , did https://github.com/scientific-python/pytest-doctestplus/pull/222 address your concern?

**Comment by pllim** [comment]
> package-data section can be removed and we can rely on setuptools-scm instead

I agree that this can be a follow-up PR. Should we open a new issue about this?

File: docs/development/testguide.rst (hunk 11 above)
Comment (astrofrog) [reply docs/development/testguide.rst:41]: They are on Mac (because it uses zsh)

Comment (WilliamJamieson) [reply docs/development/testguide.rst:41]: @astrofrog is correct, that will break on `zsh` without the `"`, while the `"` does not change anything on other shells.

File: pyproject.toml (hunk 9 above)
Comment (nstarman) [reply pyproject.toml]: We can set up repo checker made by Scientific Python. For a different PR of course.

Comment (WilliamJamieson) [reply pyproject.toml]: Yes this is my next thing to do. I worked a while ago with them to get the pre-commit hook functioning properly, but it only checks the pyproject.toml for project configuration settings.

**Comment by neutrinoceros** [comment]
Jumping in to mention the existence of [pyproject-fmt](https://pyproject-fmt.readthedocs.io/en/latest/) which comes as a pre-commit hook. I don't think it's critical important but *if* it's worth considering, this PR seems to me like the best time to enable it.

File: .flake8 [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    astropy/units/tests/test_quantity_decorator.py:F821
+    astropy/units/tests/test_quantity_typing.py:F821
+    astropy/wcs/wcs.py:F821
+    astropy/wcs/wcsapi/fitswcs.py:F821
+    astropy/wcs/wcsapi/utils.py:E127,E128
+    docs/conf.py:F401
+    examples/*.py:E1,E2,E402
+    */examples/*.py:E1,E2,E402
+
+[pycodestyle]
```
Comment (mhvk) [inline .flake8:33]: This seems weird, shouldn't it be in its own `.pycodestyle` file (or similar)?

File: docs/development/testguide.rst [hunk 14]
```diff
@@ -27,17 +27,18 @@ dependency itself, in addition to several ``pytest`` plugins that are used by
 Astropy, and will also be of general use to other packages.
 
 Since the testing dependencies are not actually required to install or use
-Astropy, they are not included in ``install_requires`` in ``setup.cfg``.
-Instead, they are listed in an ``extras_require`` section called ``test`` in
-``setup.cfg``. Developers who want to run the test suite will need to either
+Astropy, they are not in the ``pyproject.toml`` included under the ``[project]``
```
Comment (mhvk) [inline docs/development/testguide.rst:30]: Bit ungrammatical. Maybe 
```
Astropy, in the ``pyproject.toml`` file they are not included under the ``[project]``
section in ``dependencies``. Instead, they are listed under the
``[project.optional-dependences]`` section called ``test``.
Developers who...
```
```

**Review by mhvk — COMMENTED** [review]
Nice! Two small comments.

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/15247#issuecomment-1713530445

@neutrinoceros , I would prefer to revisit that as a follow-up PR as not to drag this out. Thanks!

**Comment by WilliamJamieson** [comment]
> Jumping in to mention the existence of [pyproject-fmt](https://pyproject-fmt.readthedocs.io/en/latest/) which comes as a pre-commit hook. I don't think it's critical important but _if_ it's worth considering, this PR seems to me like the best time to enable it.

I would prefer doing this, but side conversations with others have indicated to me that others are weary of adding yet more formatters. So I think it would actually be better to separate that discussion into its own PR, so it does not drag on this one.

**Comment by neutrinoceros** [comment]
no worries. I actually don't use it anywhere yet for this exact reason, so I don't intend to actively push for it !

**Comment by WilliamJamieson** [comment]
> Hello @WilliamJamieson! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:
> 
> * In the file [`astropy/coordinates/tests/test_intermediate_transformations.py`](https://github.com/astropy/astropy/blob/0adc2dfdd052d6b1ef95aa3d5d45c8de5ad1e9ce/astropy/coordinates/tests/test_intermediate_transformations.py):
> 
> > [Line 875:89](https://github.com/astropy/astropy/blob/0adc2dfdd052d6b1ef95aa3d5d45c8de5ad1e9ce/astropy/coordinates/tests/test_intermediate_transformations.py#L875): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 88 characters)

@pllim I have not seen this active in a long time in astropy. I don't think it should be as we use pre-commit and `ruff` should be catching violations like this (@nstarman is this a `ruff` bug or is it due to `ruff` setting )

**Comment by nstarman** [comment]
We don't and haven't (ever?) enforced E501

https://github.com/astropy/astropy/blob/b9b552a68038f9513616364226a4f98f2c1e0b71/.ruff.toml#L55

It's on our todo list, but it's a big job.

**Comment by astrofrog** [comment]
I thought we did enforce 100 character line lengths?

**Comment by nstarman** [comment]
AFAIK that was entirely done by human review. I can't find anything in a quick perusal of the git history.

**Comment by pllim** [comment]
Re: pep8speaks -- someone must have accidentally re-enabled it for this repo. I disabled it again but it is still enabled in the org for `astroquery` and `photutils`. FYI.

**Comment by pllim** [comment]
Back to the topic at hand, would be nice to get approval from @saimn since he had concerns before.

**Review by mhvk — APPROVED** [review]
Looks good to me. Thanks, @WilliamJamieson!

File: .pycodestyle
Comment (nstarman) [inline .pycodestyle:1]: 😢

File: astropy/utils/compat/optional_deps.py [hunk 15]
```diff
@@ -6,7 +6,7 @@
 import warnings
 
 # First, the top-level packages:
-# TODO: This list is a duplicate of the dependencies in setup.cfg "all", but
+# TODO: This list is a duplicate of the dependencies in pyproject.toml "all", but
 # some of the package names are different from the pip-install name (e.g.,
 # beautifulsoup4 -> bs4).
 _optional_deps = [
```
Comment (nstarman) [inline astropy/utils/compat/optional_deps.py:12]: In a followup PR we could handle this more automatically if python>=3.11, using `importlib.metadata`

File: pyproject.toml [hunk 16]
```diff
@@ -1,14 +1,212 @@
...
+    "objgraph",
+    "ipython>=4.2",
+    "coverage[toml]",
+    "skyfield>=1.20",
+    "sgp4>=2.3",
+]
+recommended = [
+    "scipy>=1.5",
+    "matplotlib>=3.3,!=3.4.0,!=3.5.2",
+]
```
Comment (nstarman) [inline pyproject.toml:68]: nitpick: can you move this to the top of the optional dependencies. It's the one that users are most likely to want to see.

**Review by nstarman — APPROVED** [review]
Thanks @WilliamJamieson for biting the bullet and leading this long-desired transition.
I have [dug up contentious conversations from 2021](https://github.com/astropy/astropy/pull/12276#issuecomment-953131847) asking for cfg->toml.

File: .flake8
Comment (eerovaher) [inline .flake8:1]: Why should this file exist?

Comment (nstarman) [reply .flake8:1]: I agree.
The rational is to support editors that still use `.flake8`.
But at this point I think all those same editors support ruff (can anyone find a counterexample?).
IMO `.flake8` ➡ 🗑️

Comment (nstarman) [reply .flake8:1]: But if anyone still requires `.flake8` then we can keep it around.

Comment (pllim) [reply .flake8:1]: I want it and I need it. Please keep it. Thanks!

Comment (eerovaher) [reply .flake8:1]: What do you need this for? Is this for integration with some editor and if it is then have you looked at [how to integrate Ruff with your editor](https://docs.astral.sh/ruff/editor-integrations/) instead?

Comment (pllim) [reply .flake8:1]: Yes, for my editor. It is unfair to force everyone to use ruff right away or ever. If I like flake8, please just let me use it. Does not hurt to keep this file.

File: astropy/utils/compat/optional_deps.py (hunk 15 above)
Comment (WilliamJamieson) [reply astropy/utils/compat/optional_deps.py:12]: Agreed.

File: pyproject.toml (hunk 16 above)
Comment (WilliamJamieson) [reply pyproject.toml:68]: Sure, however I think it is ordered in the same order as the current `setup.cfg`.

File: .flake8
Comment (WilliamJamieson) [reply .flake8:1]: Its been mentioned before and rejected; however, `flake8` does look in the [`tox.ini`](https://flake8.pycqa.org/en/latest/user/configuration.html#configuration-locations) for configuration so we could move it there. Indeed [`pycodestyle`](https://pycodestyle.pycqa.org/en/latest/intro.html#configuration) also allows for `tox.ini` configuration. Since this is built into both tools itself, as long as the editor is just scripting the tool it should pick up the configuration from that.

While I am not a huge fan of having these in the `tox.ini` file as they don't seem related to the file's intent, the same could be argued for there current location in the `setup.cfg`. I would advocate for this move simply so that we lessen the number of dot/configuration files in the main astropy directory.

File: .pycodestyle
Comment (WilliamJamieson) [reply .pycodestyle:1]: Agreed, see my comment on `.flake8`.

File: .flake8
Comment (WilliamJamieson) [reply .flake8:1]: Fair points on all accounts. I think continuing to support `.flake8` and `.pycodestyle` configurations should be a separate discussion. As far as moving to the `tox.ini`, the main reasoning here is that we now have a large number of configuration files for various things in the main astropy directory already. I count 15 such files at the moment, this PR replaces one of those files with two new ones. The move to the `tox.ini` is simply an option to tame this. Given PyCQA/flake8#234 and PyCQA/pycodestlye#814, I do not think we will see `pyproject.toml` configuration options for `flake8` or `pycodestyle` in the near future, which would be the ideal solution.

Comment (nstarman) [reply .flake8:1]: Yeah, https://github.com/PyCQA/flake8/issues/234 makes 0 sense. They are saying that they won't support `pyproject.toml` because it changes how a package is built. But it's not `flake8` that has the `pyproject.toml`, it's the user who's **choosing** to use a `pyproject.toml` for configuration and presumably project build info. I get that  for this reason they might want to support legacy configuration formats (aka just keep their existing code), but to dictatorially force everyone else to continue using those legacy formats is extremely irksome. Sigh. Tangent rant over.

Comment (nstarman) [reply .flake8:1]: Also, us deleting `.flake8` or `.pycodestyle` doesn't stop someone from adding it to their local repo and git-excluding the files. People are free to configure astropy however they want on local machines...

As to moving them to `tox.ini` I think another benefit beyond reducing clutter is that users *won't see the files*.  We don't have [prospector](https://pypi.org/project/prospector/) config files because we don't use it! We don't use flake8 nor pycodestyle as a *project*, only as *individuals*. I think it's misleading to advertise these configurations.

Comment (nstarman) [reply .flake8:1]: But I think this should be done in a followup PR. no point delaying this PR.

File: pyproject.toml [hunk 17]
```diff
@@ -1,14 +1,212 @@
...
+name = "astropy"
+dynamic = [
+    "version"
+]
+description = "Astronomy and astrophysics core library"
+readme = "README.rst"
+authors = [
+    { name = "The Astropy Developers", email = "astropy.team@gmail.com" }
+]
+license = { file = "LICENSE.rst" }
```
Comment (saimn) [inline pyproject.toml:11]: Would be nice to retain the SPDX identifier that was added some time ago.
```
license = { text = "BSD-3-Clause" }
```
Seems more informative. I'm not sure if we can put both in `license`, and I cannot find a clear recommendation  (https://peps.python.org/pep-0621/#license). We can also use setuptools' `license-files` to specify the file.

**Review by saimn — COMMENTED** [review]
Just a small comment, LGTM otherwise.

**Comment by saimn** [comment]
Also would be great to stop polluting PRs with unrelated discussions on config files and tools. The main concern here is packaging, and if a mistake slips in the noise we will have to deal with it when the release is out.

**Comment by nstarman** [comment]
Certainly we don't want to have a packaging problem! @saimn are you concerned about the failing weekly cron jobs?

**Comment by pllim** [comment]
> are you concerned about the failing weekly cron jobs?

Those are also being discussed separately over at #15351 . I plan to look into the mess that is devdeps separately today.

File: pyproject.toml (hunk 17 above)
Comment (WilliamJamieson) [reply pyproject.toml]: Unfortunately, we cannot do both, per PEP621 (last sentence):

> The table may have one of two keys. The file key has a string value that is a relative file path to the file which contains the license for the project. Tools MUST assume the file’s encoding is UTF-8. The text key has a string value which is the license of the project whose meaning is that of the License field from the [core metadata](https://packaging.python.org/specifications/core-metadata/). __These keys are mutually exclusive, so a tool MUST raise an error if the metadata specifies both keys.__

So I had to use the `tool.setuptools.liscense-files` to specify the actual files, see 608019c

Comment (pllim) [reply pyproject.toml]: Since @saimn 's last comment is addressed, I am going to squash and merge. Thanks, everyone!
